rbs/prism/CommentsAssociatorPrism.cc

```diff
@@ -1,17 +1,1069 @@
+            auto *call = down_cast<pm_call_node_t>(node);
+
+            if (parser.isVisibilityCall(node)) {
+                // This is a visibility modifier wrapping a method definition: `private def foo; end`
+                associateSignatureCommentsToNode(node);
+                associateAssertionCommentsToNode(node);
+                walkNode(call->receiver);
+                if (call->arguments != nullptr) {
+                    walkNodes(call->arguments->arguments);
+                }
+                consumeCommentsInsideNode(node, "call");
+            } else if (parser.isAttrAccessorCall(node)) {
+                associateSignatureCommentsToNode(node);
+                associateAssertionCommentsToNode(node);
+                walkNode(call->receiver);
+                if (call->arguments != nullptr) {
+                    walkNodes(call->arguments->arguments);
+                }
+                consumeCommentsInsideNode(node, "call");
+            } else if (call->arguments != nullptr && call->arguments->arguments.size == 1 &&
+                       parser.isSafeNavigationCall(node) && parser.isSetterCall(node)) {
+                // Handle safe navigation setter calls: `foo&.bar = val #: Type`
+                associateAssertionCommentsToNode(call->arguments->arguments.nodes[0]);
+                walkNode(call->arguments->arguments.nodes[0]);
+                consumeCommentsInsideNode(node, "safe navigation call");
+            } else if (parser.resolveConstant(call->name) == "[]=" || parser.isSetterCall(node)) {
+                // This is an assign through a call, either: `foo[key]=(y)` or `foo.x=(y)`
+                //
+                // Note: the parser groups the args on the right hand side of the assignment into an array node:
+                //  * for `foo.x = 1, 2` the args are `[1, 2]`
+                //  * for `foo[k1, k2] = 1, 2` the args are `[k1, k2, [1, 2]]`
+                //
+                // We always apply the cast starting from the last arg by walking them in reverse order.
+                if (call->arguments != nullptr) {
+                    for (int i = call->arguments->arguments.size - 1; i >= 0; i--) {
+                        walkNode(call->arguments->arguments.nodes[i]);
+                    }
+                }
+                walkNode(call->receiver);
+                consumeCommentsInsideNode(node, "call");
+            } else {
+                associateAssertionCommentsToNode(node);
+                walkNode(call->receiver);
+                if (call->arguments != nullptr) {
+                    walkNodes(call->arguments->arguments);
+                }
+                walkNode(call->block);
+            }
+            break;
+        }
+        case PM_DEF_NODE: {
+            auto *def = down_cast<pm_def_node_t>(node);
+            auto defKeywordLoc = translateLocation(def->def_keyword_loc);
+            contextAllowingTypeAlias.push_back(make_pair(false, defKeywordLoc));
```

Comment:
With these `push_back` + `pop_back()` pairs, is there any risk that an exception in the middle messes up your stack?

If yes, you might want to extract a helper that takes a lambda that does the stuff in the middle, with a `try`/`finally` to ensure you always pop